python/lsst/daf/butler/core/dimensions/sets.py

```diff
@@ -0,0 +1,402 @@
+        **kwds : unspecified
+            Additional subclass-specific arguments.
+
+        Returns
+        -------
+        result : instance of a subclass of `DimensionSet`
+            A new set of the same type as ``self`` containing any elements
+            in all input sets.
+        """
+        names = set(self.names)
+        for other in others:
+            UniverseMismatchError.check(self, other)
+            names &= other.names
+        return type(self)(self.universe, names, **kwds)
+
+    def __or__(self, other):
+        return self.union(other)
+
+    def __and__(self, other):
+        return self.intersection(other)
+
+    def __xor__(self, other):
+        return self.symmetric_difference(other)
+
+    def __sub__(self, other):
+        return self.difference(other)
+
+    def copy(self):
+        """Return a shallow copy of this set.
+
+        Because `Dimension` instances are immutable, there should be no need
+        to ever perform a deep copy.
+        """
+        result = type(self)(self.universe)
+        result._elements = self._elements.copy()
+        return result
+
+    @abstractmethod
+    def _populate(self, names, **kwds):
+        r"""Extract the `Dimension`\s with the given names from
+        ``self.universe`` (in order) and add them to ``self._elements``.
+
+        Must be implemented by subclasses, which may opt to also include
+        other `Dimension`\s not in ``names``.
+
+        Additional keyword arguments are forwarded directly by ``union``,
+        ``intersection``, and the class constructor.
+        Subclasses must support at least the case where no additional keywords
+        arguments are provided.
+        """
+        raise NotImplementedError()
+
+
+class DimensionUnitSet(DimensionSet):
+    r"""A concrete `DimensionSet` for `DimensionUnit` instances.
+
+    `DimensionUnitSet`\s always include all required dependencies of all
+    of the `DimensionUnit`\s they contain, and are sorted topologically
+    (with an unspecified but guaranteed deterministic ordering to break
+    topological ties).
+
+    Automatic expansion to include required dependencies means that set
+    operations that remove elements (`symmetric_difference` and `difference`)
+    may not work as naively expected; the direct operation may remove a
+    `DimensionUnit` from the result while retaining one that depends on it,
+    which would then cause it to be added back into the result set.
+
+    The constructor and `union` and `intersection` methods for
+    `DimensionUnitSet` take an extra ``optional`` argument that controls
+    whether optional dependencies of all elements are also included in
+    the returned set.  This defaults to `False`.
+    """
+
+    def __repr__(self):
+        return f"DimensionUnitSet({str(self)})"
+
+    def _populate(self, names, optional=False):
+        expanded = set()
+
+        # recursive local function to add dependencies of dependencies
+        def visit(unit):
```

Comment:
I'd probably use generator instead of closure for this.

Reply:
When you get a chance, could you elaborate?  I'm not quite sure what you're proposing.

Reply:
I mean something like:


            def visit(unit):
                yield unit.name
                for dep in unit.dependencies(optional=optional):
                    yield from visit(dep)

            expaned = set(visit(unit))